Add firehose reloaded

max-standard opened the pull request:
This pull request does the following:

1. Add _[firehose_config](https://github.com/cloudchacho/terraform-google-hedwig-topic/pull/2/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR5-R6)_ setting which let's you _firehose_—read as, _save_—Hedwig messages to Google Cloud Storage (GCS). This implementation is based on the excellent [_cloud_storage_config_](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/pubsub_subscription#cloud_storage_config) setting—added in google provider version [_4.78.0_](https://github.com/hashicorp/terraform-provider-google/releases/tag/v4.78.0)—which creates a _[Pub/Sub Cloud Storage](https://cloud.google.com/pubsub/docs/create-cloudstorage-subscription)_ subscription and obviates Dataflow configuration tasks.
2. Remove Dataflow configs.

max-standard committed:
Add firehose reloaded

Based on https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/pubsub_subscription#example-usage---pubsub-subscription-push-cloudstorage example.

maroux commented on main.tf:
```diff
@@ -2,7 +2,22 @@ resource "google_pubsub_topic" "topic" {
+  cloud_storage_config {
+    bucket          = var.firehose_bucket
+    filename_prefix = local.firehose_prefix
+    max_bytes       = 1000
```
I'd make this and the next one variables to play around with optimal settings.

max-standard replied:
Sounds good; also see https://github.com/cloudchacho/terraform-google-hedwig-topic/pull/2/commits/b5a0aa376d1ee0506a8180ae821d0eda09d07b9e which makes _[avro_config](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/pubsub_subscription#avro_config)_ setting configurable since I recall Hedwig can use the simpler JSON format too.

max-standard committed:
Remove dataflow

Because firehose reloaded uses cloud_storage_config subscription setting instead.

max-standard committed:
Write messages in Avro format

max-standard committed:
Make max_duration and max_bytes configurable

max-standard committed:
Add dynamic avro_config

shsamkit commented on main.tf:
```diff
@@ -2,7 +2,28 @@ resource "google_pubsub_topic" "topic" {
+
+  cloud_storage_config {
+    bucket          = var.firehose_bucket
+    filename_prefix = local.firehose_prefix
```
Q.  What will be the final name of the files? 
Context: I see there's a `filename_suffix` as well, which I figure is for denfining filenames but I am not sure what the end result would look like. I assume it will create multiple files for each subscription.

shsamkit replied:
Nvm, found the info : https://cloud.google.com/pubsub/docs/create-cloudstorage-subscription#file_names

max-standard replied:
That's right and I added the _`filename_suffix`_ setting to help with tasks like adding an _`.avro`_ suffix to filenames.

shsamkit commented on variables.tf:
```diff
@@ -3,39 +3,40 @@ variable "topic" {
 
-variable "dataflow_tmp_gcs_location" {
-  description = "A gs bucket location for storing temporary files by Google Dataflow, e.g. gs://myBucket/tmp"
+variable "firehose_bucket" {
```
Any specific reason to prefer premitive type over an object? An object type will help with readability, and you will not need to definte `default` values for required fields like `firehose_bucket`

max-standard replied:
Sounds good, I added a _`firehose_config`_ object to contain those settings and improve readability—see _[c5c0199](https://github.com/cloudchacho/terraform-google-hedwig-topic/commit/c5c0199aaf000db9aee6463f2e5d7aee7fe52b0e?diff=split&w=0#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR5-R10)_ commit. **Edit:** **Note:** I didn't do this intentionally; but I realized that I used the new _Optional attributes for object type constraints_ Terraform feature—see _[terraform v1.3.0](https://github.com/hashicorp/terraform/releases/tag/v1.3.0)_ release notes.

max-standard committed:
Bump google provider to 4.78.0

because 4.78.0 added cloud_storage_config google_pubsub_subscription block--see https://github.com/hashicorp/terraform-provider-google/releases/tag/v4.78.0 page.

max-standard committed:
Add firehose_config object

Addresses https://github.com/cloudchacho/terraform-google-hedwig-topic/pull/2#discussion_r1417627468 comment.

max-standard committed:
Move ^hedwig- topic naming convention to accurate context

max-standard committed:
Add firehose_config docs

max-standard committed:
nit: docs: Remove trailing space and stylize GitHub

max-standard committed:
Make firehose_config optional

Gotcha: You can't actually set bucket to null; but this is just to express the fact that firehose_config variable is optional. If you do use firehose_config variable, then bucket is required.

max-standard committed:
Improve firehose_config default handling

1. If you set firehose_config to null, then apply defaults
2. If you set firehose_config.enabled but not bucket, then fail plan

max-standard committed:
Remove <topic> hack

because calling modules can set the filename_prefix to topic name if desired—that is, the <topic> hack was redundant.

maroux commented on variables.tf:
```diff
@@ -2,40 +2,34 @@ variable "topic" {
+  type = object({
+    # Declares the bucket to firehose messages to.
+    # Gotcha: Your bucket should already exist with complicated IAM permissions—see https://cloud.google.com/pubsub/docs/create-cloudstorage-subscription#assign_roles_cloudstorage page.
+    # Gotcha: You can't actually set bucket to null; but this null default is just to express the fact that firehose_config variable is optional and if you do set firehose_config.enabled = true; but don't set bucket, then the plan will fail.
```
Hm why not make this object optional so the presence of it signifies whether firehose is enabled or not? That way you don't need this gotcha or the "enabled" field?

max-standard replied:
The _firehose_config_ object _is_ optional because I set _[default = {}](https://github.com/cloudchacho/terraform-google-hedwig-topic/pull/2/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR31)_. I added the _enabled_ attribute because I guessed toggling _enabled_ was useful (rather than: when you want to disable firehose, then comment out or remove _firehose_config_ input).

I can remove _enabled_ and make the check work like (change incoming):
```
resource "google_pubsub_subscription" "firehose" {
  count = var.firehose_config != null ? 1 : 0
# ...
```
But guess: the bucket gotcha will remain since I'll need to do a default like:
```
variable "firehose_config" {
  type = object({
    bucket = string
    # ...
    })
  default = {
    # Gotcha: bucket attribute can't actually be null; this is just to make firehose_config default work
    bucket = null
  }
}
```

maroux replied:
Yeah I like strongly typed objects as much as feasible and this gotcha can be avoided. Would making default as null solve your problem?

max-standard replied:
That works—works around the bucket gotcha! And also achieves the following:
1. Make firehose_config object _optional_ by defaulting to null.
1. Declare the bucket as a _required_ string field.

Guess: I must have overlooked that possibility where you default your object the null; but the required field still won't error—surprise; but it works—see https://github.com/cloudchacho/terraform-google-hedwig-topic/pull/2/commits/bc4a5c94f46cc57fb4a88fad4e57a95fade374eb commit.

max-standard committed:
Set firehose_config default to null

Note: Apply https://github.com/cloudchacho/terraform-google-hedwig-topic/pull/2#discussion_r1427304819 comment to avoid bucket null default gotcha.

max-standard committed:
Only check var.firehose_config != null once

max-standard committed:
Reference google_pubsub_topic.topic.id

This creates the correct topic<-subscription(cloud_storage) dependency.

max-standard committed:
Add explicit bucket dependency instead of only a comment
